cli/server/dump_bin_put.go

```diff
@@ -0,0 +1,364 @@
+		log.Println("Fetching block", index)
+	}
+	block, err := rpcClient.GetBlockByIndex(uint32(index))
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch block %d: %w", index, err)
+	}
+
+	var buf bytes.Buffer
+	bw := io.NewBinWriterFromIO(&buf)
+
+	block.EncodeBinary(bw)
+
+	if bw.Err != nil {
+		return nil, fmt.Errorf("failed to encode block %d: %w", index, bw.Err)
+	}
+
+	return buf.Bytes(), nil
+}
+
+func uploadBObjWithSlicer(ctx context.Context, clientSDK *client.Client, account wallet.Account, containerID cid.ID, objData []byte, attributeKey, attributeValue string, sessionToken *session.Object) error {
+	signer := user.NewAutoIDSignerRFC6979(account.PrivateKey().PrivateKey)
+	var ownerID user.ID
+	ownerID.SetScriptHash(account.PrivateKey().GetScriptHash())
+
+	slc, err := slicer.New(ctx, clientSDK, signer, containerID, ownerID, sessionToken)
+	if err != nil {
+		return fmt.Errorf("failed to create slicer: %w", err)
+	}
+
+	attrs := []object.Attribute{
+		*object.NewAttribute(attributeKey, attributeValue),
+	}
+
+	_, err = slc.Put(ctx, bytes.NewReader(objData), attrs)
+	if err != nil {
+		return fmt.Errorf("failed to slice and upload block data: %w", err)
+	}
+
+	return nil
+}
+
+// updateIndexFiles updates the index files in the container.
+func updateIndexFiles(ctx *cli.Context, clientSDK *client.Client, containerID cid.ID, account wallet.Account, currentHeight uint, attributeKey string, blockAttributeKey string) error {
+	log.Println("Updating index files...")
+
+	prm := client.PrmObjectSearch{}
+	filters := object.NewSearchFilters()
+	filters.AddFilter(attributeKey, fmt.Sprintf("%d", 0), object.MatchNumGE)
+	prm.SetFilters(filters)
+
+	objectIDs, err := neofs.ObjectSearch(ctx.Context, clientSDK, account.PrivateKey(), containerID.String(), prm)
+	if err != nil {
+		return fmt.Errorf("no OIDs found for index files: %w", err)
+	}
+
+	existingIndexCount := uint(len(objectIDs))
+	expectedIndexCount := currentHeight / uint(indexFileSize)
+
+	if existingIndexCount == expectedIndexCount {
+		return nil
+	}
+
+	var buf bytes.Buffer
+	fileIndex := int(existingIndexCount)
+
+	ctxWithCancel, cancel := context.WithCancel(ctx.Context)
+	defer cancel()
+
+	workerPool := make(chan struct{}, maxParallelSearches)
+	errCh := make(chan error, 1)
+	wg := sync.WaitGroup{}
+
+	for i := existingIndexCount; i < expectedIndexCount; i++ {
+		startIndex := i * indexFileSize
+		endIndex := startIndex + indexFileSize
+		if endIndex > currentHeight {
+			return fmt.Errorf("end index %d exceeds current height %d", endIndex, currentHeight)
+		}
+
+		blockOids := make([]oid.ID, indexFileSize)
+		for j := startIndex; j < endIndex; j++ {
+			select {
+			case <-ctxWithCancel.Done():
+				return <-errCh
+			default:
+			}
+
+			wg.Add(1)
+			workerPool <- struct{}{}
+
+			go func(index uint) {
+				defer wg.Done()
+				defer func() { <-workerPool }()
+
+				oidBlock, err := searchBlockOID(ctx, clientSDK, account, containerID, blockAttributeKey, index)
```

Comment:
You perform Search request for every block, it's very inefficient. We need to construct index file, we may do the following:

1. Perform a single Search request, search for 128K objects at once.
2. Allocate buffer of `128K * OIDSize` bytes
3. Iterate through the received object headers, for every object: get block index attribute from the header and store block object OID in the buffer by given index.
4. Store given buffer as an index object.

Reply:
With your approach instead of 128k searches for every block, we will have 1 big search + 128k getHeader requests + 128k parse for each header. It is because search request returns only oids.

Reply:
>  It seems that 

Have you tried? If not, then let's try. Search is not cheap.